### PR TITLE
Add TF notebook for TPU

### PR DESCRIPTION
**Pull request by Rocketknight1**
This PR adds a TPU training notebook for TF, with several sections covering different ways to build your input pipeline. It's designed to be the hands-on/code-focused partner to the tutorial doc (I'll make that PR next!), which will be more high-level and not purely focused on code you can execute right now.

This ended up touching on a lot of different areas and being a bit more work to write than I thought, so I tagged a couple of extra people to review here - @amyeroberts to make sure I didn't butcher your image code too badly, and @sayakpaul to make sure my `tf.data` code isn't going to offend the TF community too badly. If you're short on time, don't worry about doing a full review of everything!

### CODE REVIEW

**Comment by Rocketknight1** [comment]
Tutorial doc PR is up at https://github.com/huggingface/transformers/pull/21429

File: examples/tpu_training-tf.ipynb [hunk 1]
```diff
@@ -0,0 +1,904 @@
+{
```
Comment (sayakpaul) [inline examples/tpu_training-tf.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #10.    # For simplicity, let's just tokenize our dataset as NumPy arrays 
</pre><p>What are other options here? Do you want to spend a sentence to make our users aware of that? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/313/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (sayakpaul) [inline examples/tpu_training-tf.ipynb:1]: <p><code>BATCH_SIZE = local_batch_size * strategy.num_replicas_in_sync</code> is the recommended way :)  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/313/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (sayakpaul) [inline examples/tpu_training-tf.ipynb:1]: <p>Love it!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/313/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (sayakpaul) [inline examples/tpu_training-tf.ipynb:1]: <p><code>tf.numpy_function</code> can work on a TPU VM? My god 🤯 </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/313/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

Comment (sayakpaul) [inline examples/tpu_training-tf.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #14.    with strategy.scope():
</pre><p> </p><p>Nit: Sometimes, you use 4 spaces and sometimes 2. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/313/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='30'/>

Comment (sayakpaul) [inline examples/tpu_training-tf.ipynb:1]: <blockquote>more work, and often requires you to pay for cloud storage, but it works on all kinds of TPU instances, and allows for truly massive data pipeline throughput.&nbsp;</blockquote><p> </p><p>Works on all kinds of devices (CPU, GPU, TPU). </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/313/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='37'/>

Comment (sayakpaul) [inline examples/tpu_training-tf.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    def decode_fn(sample):
</pre><p>Two spaces. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/313/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='45'/>

Comment (sayakpaul) [inline examples/tpu_training-tf.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #12.      tf_dataset = tf_dataset.shuffle(len(dataset)).batch(BATCH_SIZE, drop_remainder=True)
</pre><p>Can spend one sentence on why <code>drop_remainder</code> is set to True.  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/313/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='45'/>

Comment (sayakpaul) [inline examples/tpu_training-tf.ipynb:1]: <p>Help me understand this. </p><p> </p><p>Since the dataset can fit into memory (and therefore the converted TFRecord), <code>model.fit(tf_dataset)</code> works here without errors on Colab TPUs?  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/313/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='47'/>

**Review by sayakpaul — APPROVED** [review]
Awesome work! Really liked all the different scenarios and how they were discussed.

File: examples/tpu_training-tf.ipynb (hunk 1 above)
Comment (amyeroberts) [reply examples/tpu_training-tf.ipynb:1]: <p>"Hugging Face" as two words again 🤗</p>

Comment (amyeroberts) [inline examples/tpu_training-tf.ipynb:1]: <p>Same here re "Hugging Face"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/313/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='35'/>

Comment (amyeroberts) [inline examples/tpu_training-tf.ipynb:1]: <p>Same here - "Hugging Face"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/313/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='42'/>

Comment (amyeroberts) [inline examples/tpu_training-tf.ipynb:1]: <p>Same here - "Hugging Face" 🤗</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/313/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='50'/>

Comment (amyeroberts) [inline examples/tpu_training-tf.ipynb:1]: <p>This is a nice summary :) </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/313/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='59'/>

**Review by amyeroberts — APPROVED** [review]
Nice - very clear and easy to follow. Thanks for adding this! 

Only comment is about replacing "HuggingFace" with "Hugging Face" throughout

File: examples/tpu_training-tf.ipynb (hunk 1 above)
Comment (sgugger) [inline examples/tpu_training-tf.ipynb:1]: <p>Insist on the fact that you need a fixed length for TPUs, which is why you pad to max length.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/313/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (sgugger) [inline examples/tpu_training-tf.ipynb:1]: <p>No 2-line indentation. This is just evil.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/313/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (sgugger) [inline examples/tpu_training-tf.ipynb:1]: <p>4-line indent.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/313/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='32'/>

Comment (sgugger) [inline examples/tpu_training-tf.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #14.    with strategy.scope():
</pre><p>4-line indent again.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/313/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='41'/>

Comment (sgugger) [inline examples/tpu_training-tf.ipynb:1]: <p>4-lines indent.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/313/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='45'/>

Comment (sgugger) [inline examples/tpu_training-tf.ipynb:1]: <p>4-lines indent</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/313/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='58'/>

Comment (sgugger) [inline examples/tpu_training-tf.ipynb:1]: <p>4-lines indent.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/313/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='62'/>

Comment (sgugger) [inline examples/tpu_training-tf.ipynb:1]: <p>4-lines indent.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/313/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='64'/>

**Review by sgugger — COMMENTED** [review]
Thanks for drafting this notebook. I am sorry to see that you have fallen victim to Colab's evil propaganda towards 2-space indentation. We shall forever resist against the oppressor and let the 4-space indentation shine upon us.

**Review by gante — APPROVED** [review]
LGTM 💛 

(except for the 2-character indentation dark plot to dominate the world, that it)

File: examples/tpu_training-tf.ipynb (hunk 1 above)
Comment (Rocketknight1) [reply examples/tpu_training-tf.ipynb:1]: <p>I think variable lengths are possible, at the cost of a new XLA compilation each time! (This is definitely the case with standard XLA in TF, but I haven't benchmarked it extensively with TPUs)</p><p> </p>

Comment (sgugger) [reply examples/tpu_training-tf.ipynb:1]: Yes, it recompiles each time, which makes it unusable in practice ;-)

**Comment by Rocketknight1** [comment]
Fixed all the indentation and HuggingFace spelling!

**Comment by sayakpaul** [comment]
Merge away!

**Review by sgugger — APPROVED** [review]

File: examples/tpu_training-tf.ipynb (hunk 1 above)
Comment (Rocketknight1) [reply examples/tpu_training-tf.ipynb:1]: <p>In our other TF-XLA notebooks we recommend using <code>pad_to_multiple_of</code> with a multiple like 32 or 64 - this means there's only a few unique shapes, but you still get a big performance boost compared to max length padding or running without XLA. I'll check if that works on TPU too!</p><p> </p>

Comment (sayakpaul) [reply examples/tpu_training-tf.ipynb:1]: > In our other TF-XLA notebooks we recommend using pad_to_multiple_of with a multiple like 32 or 64 - this means there's only a few unique shapes, but you still get a big performance boost compared to max length padding or running without XLA. I'll check if that works on TPU too!

You mean checking if shorter sequence lengths like 32 or 64 help boost the speed? @Rocketknight1

Comment (Rocketknight1) [reply examples/tpu_training-tf.ipynb:1]: <p>Done! (Here and in all the other places we define BATCH_SIZE)</p><p> </p>

Comment (Rocketknight1) [reply examples/tpu_training-tf.ipynb:1]: <p>I added a comment that we discuss other options for creating the dataset below.</p>

Comment (Rocketknight1) [reply examples/tpu_training-tf.ipynb:1]: <p>Spelling fixed!</p>

Comment (Rocketknight1) [reply examples/tpu_training-tf.ipynb:1]: <p>Fixed!</p>

Comment (Rocketknight1) [reply examples/tpu_training-tf.ipynb:1]: <p>I added this in the first place where drop_remainder is used!</p>

Comment (Rocketknight1) [reply examples/tpu_training-tf.ipynb:1]: <p>The dataset could fit into memory, but in this example we save it as TFRecord files on the local disk of the VM, and as a result the TPU node cannot read it. If you try running this code on Colab, you get this error:</p><p> </p><p><code>UNIMPLEMENTED: {{function_node __inference_train_function_24666}} File system scheme '[local]' not implemented (file: 'dataset.tfrecords')</code></p>

Comment (Rocketknight1) [reply examples/tpu_training-tf.ipynb:1]: <p>Yes, but also that XLA compilation and cacheing works the same way. </p><p> </p><p>When running TF normally, multiple XLA compilations can be cached. For example, if you pad to a multiple of 64 and so your dataset only contains three unique batch shapes, for example (8, 128), (8, 192) and (8, 256), then TF will run an XLA compilation for each of those shapes the first time it encounters them, but after that it will use the cached XLA programs and training will be fast.</p><p> </p><p>I'm going to check if the same thing works on TPU!</p>

Comment (sayakpaul) [reply examples/tpu_training-tf.ipynb:1]: <p>Maybe throw in another comment to explain, this code will run perfectly fine on a CPU or a TPU VM but not on a Colab TPU throwing the error you snippeted (I just invented this term). </p>

Comment (Rocketknight1) [reply examples/tpu_training-tf.ipynb:1]: <p>Just tested it and confirmed it works the same way!</p>

Comment (Rocketknight1) [reply examples/tpu_training-tf.ipynb:1]: <p>I'll explain this in more detail in the XLA section of the tutorial doc, which will be linked once it's uploaded.</p>

Comment (Rocketknight1) [reply examples/tpu_training-tf.ipynb:1]: <p>Done!</p>

**Comment by Rocketknight1** [comment]
All comments were addressed, so I merged this - I'm going to avoid promoting it until the tutorial is up too and the links between them are added!
